subprojects/core/src/main/java/org/gradle/api/tasks/GradleBuild.java

```diff
@@ -190,24 +119,98 @@ public void setTasks(Collection<String> tasks) {
      * @since 6.0
      */
     @Internal
-    @ToBeReplacedByLazyProperty
-    public String getBuildName() {
-        return buildName;
-    }
-
-    /**
-     * Sets build name to use for the nested build.
-     *
-     * @param buildName the build name to use for the nested build
-     * @since 6.0
-     */
-    public void setBuildName(String buildName) {
-        this.buildName = buildName;
-    }
+    @Optional
+    @ReplacesEagerProperty
+    public abstract Property<String> getBuildName();
 
     @TaskAction
     void build() {
         // TODO: Allow us to inject plugins into nested builds too.
-        runNestedRootBuild(buildName, (StartParameterInternal) getStartParameter(), getServices());
+        runNestedRootBuild(getBuildName().getOrNull(), (StartParameterInternal) getStartParameter(), getServices());
+    }
+
+    static class StartParameterAdapter {
+        /**
+         * Returns the full set of parameters that will be used to execute the build.
+         *
+         * @return the parameters. Never returns null.
+         */
+        @BytecodeUpgrade
+        static StartParameter getStartParameter(GradleBuild task) {
+            return task.getStartParameter().get();
+        }
+
+        /**
+         * Sets the full set of parameters that will be used to execute the build.
+         *
+         * @param startParameter the parameters. Should not be null.
+         */
+        @BytecodeUpgrade
+        static void setStartParameter(GradleBuild task, StartParameter startParameter) {
+            task.getStartParameter().set(startParameter);
+        }
+
+        /**
+         * Returns the project directory for the build. Defaults to the project directory.
+         *
+         * @return The project directory. Never returns null.
+         */
+        @Internal
+        @BytecodeUpgrade
+        static File getDir(GradleBuild task) {
+            return task.getStartParameter().get().getCurrentDir();
+        }
+
+        /**
+         * Sets the project directory for the build.
+         *
+         * @param dir The project directory. Should not be null.
+         * @since 4.0
+         */
+        @BytecodeUpgrade
+        static void setDir(GradleBuild task, File dir) {
```

Comment:
@asodja can it work that way? It fails in the tests and I'm not sure why

Reply:
```
Could not set unknown property 'dir' for task
```
and
```
Cannot set the value of read-only property 'tasks'
```

Reply:
The problem is, that setters setDir and setTasks are removed, but they are used in tests. And due to that you then have source incompatible changes, which we cannot upgrade via bytecode upgrdades.

So in that case we should have a property like `DIrectoryProperty getDir()`. The question is, what priority does it get then vs StartParameter.

Reply:
What I would do is to leave StartParameter as normal property, since it's actually a `@Nested` and we decided not to migrate them (see https://github.com/gradle/gradle/issues/25838).

Then I would migrate `getTasks()` and `getDir` to properties.

And then in `getStartParameter()` we could do:
```
  startParameter.setCurrentDir(getDir().getAsFile().getOrNull());
  startParameter.setTaskNames(getTasks().get());
  return startParameter;
```

Reply:
>  to leave StartParameter as normal property

That is similar to what I [proposed](https://github.com/gradle/gradle/pull/29963#discussion_r1683057760) IIUC. 

I think the issue here is that we would need to decide which properties to add to cover the use cases since startParameter has a lot of them, and `getStartParameter().setSomething` is a thing.